File: src/main/java/com/leijendary/spring/iamtemplate/event/producer/NotificationProducer.java
package com.leijendary.spring.iamtemplate.event.producer;

import com.leijendary.spring.iamtemplate.event.schema.NotificationSchema;
import org.springframework.context.annotation.Bean;
import org.springframework.messaging.Message;
import org.springframework.stereotype.Component;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Sinks;

import java.util.function.Supplier;

import static reactor.core.publisher.Sinks.many;

@Component
public class NotificationProducer extends AbstractProducer<NotificationSchema> {

    private final Sinks.Many<Message<NotificationSchema>> smsBuffer = many().multicast().onBackpressureBuffer();
    private final Sinks.Many<Message<NotificationSchema>> emailBuffer = many().multicast().onBackpressureBuffer();

    @Bean
    public Supplier<Flux<Message<NotificationSchema>>> notificationSms() {
        return smsBuffer::asFlux;
    }

    @Bean
    public Supplier<Flux<Message<NotificationSchema>>> notificationEmail() {
        return emailBuffer::asFlux;
    }

    public void sms(final NotificationSchema notificationSchema) {
        final var message = message(notificationSchema);

        smsBuffer.tryEmitNext(message);
    }

    public void email(final NotificationSchema notificationSchema) {
        final var message = message(notificationSchema);

        emailBuffer.tryEmitNext(message);
    }
}

File: src/main/kotlin/com/leijendary/spring/template/iam/message/NotificationProducer.kt
package com.leijendary.spring.template.iam.message

import com.leijendary.spring.template.iam.core.config.properties.KafkaTopicProperties
import com.leijendary.spring.template.iam.core.extension.toJson
import com.leijendary.spring.template.iam.message.Topic.NOTIFICATION_EMAIL
import com.leijendary.spring.template.iam.message.Topic.NOTIFICATION_PUSH
import com.leijendary.spring.template.iam.message.Topic.NOTIFICATION_SMS
import com.leijendary.spring.template.iam.model.EmailMessage
import com.leijendary.spring.template.iam.model.PushMessage
import com.leijendary.spring.template.iam.model.SmsMessage
import org.springframework.kafka.core.KafkaTemplate
import org.springframework.retry.annotation.Retryable
import org.springframework.stereotype.Component

@Component
class NotificationProducer(
    private val kafkaTemplate: KafkaTemplate<String, String>,
    private val kafkaTopicProperties: KafkaTopicProperties
) {
    @Retryable
    fun email(emailMessage: EmailMessage) {
        val topic = kafkaTopicProperties.nameOf(NOTIFICATION_EMAIL)

        kafkaTemplate.send(topic, emailMessage.toJson())
    }

    @Retryable
    fun push(pushMessage: PushMessage) {
        val topic = kafkaTopicProperties.nameOf(NOTIFICATION_PUSH)

        kafkaTemplate.send(topic, pushMessage.toJson())
    }

    @Retryable
    fun sms(smsMessage: SmsMessage) {
        val topic = kafkaTopicProperties.nameOf(NOTIFICATION_SMS)

        kafkaTemplate.send(topic, smsMessage.toJson())
    }
}
